### PR TITLE
Initial implementation of PrivateConstructorChecker

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
@nikitin-da PTAL!

### CODE REVIEW

File: private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                    .check();
+
+            fail();
+        } catch (AssertionError expected) {
+            assertEquals("Constructor must be private", expected.getMessage());
+        }
+    }
+
+    static class ClassWithProtectedConstructor {
+        ClassWithProtectedConstructor() {
```
Comment (nikitin-da) [inline private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: protected? =)

File: private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                    .check();
+
+            fail();
+        } catch (AssertionError expected) {
+            assertEquals("Constructor must be private", expected.getMessage());
+        }
+    }
+
+    static class ClassWithPublicConstructor {
+        ClassWithPublicConstructor() {
```
Comment (nikitin-da) [inline private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Public

Comment (artem-zinnatullin) [inline private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: SHIT

File: private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java (hunk 1 above)
Comment (artem-zinnatullin) [inline private-constructor-checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: SHIT SHIT SHIT

File: private-constructor-checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+            final Throwable cause = e.getCause();
+
+            // It's okay case if we expect some exception from this constructor
+            if (expectedTypeOfException != null) {
+                if (!expectedTypeOfException.equals(cause.getClass())) {
+                    throw new IllegalStateException("Expected exception of type = "
+                            + expectedTypeOfException + ", but was exception of type = "
+                            + e
```
Comment (nikitin-da) [inline private-constructor-checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: We just need cause here

File: private-constructor-checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    );
+                }
+
+                if (expectedExceptionMessage != null) {
+                    if (!expectedExceptionMessage.equals(cause.getMessage())) {
+                        throw new IllegalStateException("Expected exception message = '"
+                                + expectedExceptionMessage + "', but was = '"
+                                + cause.getMessage() + "'",
+                                e
```
Comment (nikitin-da) [inline private-constructor-checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: And here

**Comment by artem-zinnatullin** [comment]
@nikitin-da I've fixed all comments :)

**Comment by nikitin-da** [comment]
:+1:

**Pull request merged**